tensorzero_internal/src/model.rs

```diff
@@ -194,6 +290,41 @@ impl ModelConfig {
     }
 }
 
+async fn stream_with_cache_write(
+    raw_request: String,
+    model_request: ModelProviderRequest<'_>,
+    clients: &InferenceClients<'_>,
+    mut stream: ProviderInferenceResponseStream,
+) -> Result<ProviderInferenceResponseStream, Error> {
+    let cache_key = model_request.get_cache_key()?;
+    let clickhouse_info = clients.clickhouse_connection_info.clone();
+    Ok(Box::pin(async_stream::stream! {
+        let mut buffer = vec![];
+        let mut errored = false;
+        while let Some(chunk) = stream.next().await {
+            if !errored {
+                match chunk.as_ref() {
+                    Ok(chunk) => {
+                        buffer.push(chunk.clone());
+                    }
+                    Err(e) => {
+                        tracing::warn!("Skipping cache write for stream response due to error in stream: {e}");
+                        errored = true;
+                    }
+                }
+            }
+            yield chunk;
+        }
+        if !errored {
+            let _ = start_cache_write_streaming(
+                &clickhouse_info,
+                cache_key,
+                buffer,
+                &raw_request,
+            );
+        }
```

Comment:
does this end up as a trailing write? maybe add a comment clarifying/

Reply:
I added a comment to `start_cache_write_streaming`